File: web/web/navigation/update.php
<?php
/*
Copyright (©) 2003-2013 Teus Benschop.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.
 
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
  
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


require_once ("../bootstrap/bootstrap.php");
page_access_level (CONSULTANT_LEVEL);


// The navigator does not fetch the active Bible from the user configuration,
// but from the query generated in the webbrowser by Javascript.
// The reason is that the Navigator can be independent of any Bible.
// If there's no Bible, the Navigator has all the 66 canonical books, plus their chapters and verses.
// If the Navigator has a Bible, it uses the books / chapters / verses from that Bible.
@$bible = $_GET['bible'];


@$goback = $_GET['goback'];
if (isset ($goback)) {
  Navigation_Logic::goBack ();
}


@$goforward = $_GET['goforward'];
if (isset ($goforward)) {
  Navigation_Logic::goForward ();
}


@$getbibles = $_GET['getbibles'];
if (isset ($getbibles)) {
  echo Navigation_Logic::getBiblesFragment ($bible);
  die;
}


@$setbible = $_GET['setbible'];
if (isset ($setbible)) {
  $database_config_user = Database_Config_User::getInstance ();
  $database_config_user->setBible ($bible);
}


@$getbooks = $_GET['getbooks'];
if (isset ($getbooks)) {
  echo Navigation_Logic::getBooksFragment ($bible);
  die;
}


@$setbook = $_GET['setbook'];
if (isset ($setbook)) {
  Navigation_Logic::setBook ($setbook);
}


@$getchapters = $_GET['getchapters'];
if (isset ($getchapters)) {
  echo Navigation_Logic::getChaptersFragment ($bible);
  die;
}


@$setchapter = $_GET['setchapter'];
if (isset ($setchapter)) {
  Navigation_Logic::setChapter ($setchapter);
}


@$getverses = $_GET['getverses'];
if (isset ($getverses)) {
  echo Navigation_Logic::getVersesFragment ($bible);
  die;
}


@$setverse = $_GET['setverse'];
if (isset ($setverse)) {
  Navigation_Logic::setVerse ($setverse);
}


@$getentry = $_GET['getentry'];
if (isset ($getentry)) {
  echo Navigation_Logic::getEntry ();
  die;
}


@$passage = $_GET['passage'];
if (isset ($passage)) {
  Navigation_Logic::setPassage ($bible, $passage);
  die;
}


echo Navigation_Logic::getNavigator ($bible);


?>

File: windows/webroot/navigation/update.cpp
/*
 Copyright (©) 2003-2015 Teus Benschop.
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <navigation/update.h>
#include <filter/roles.h>
#include <filter/string.h>
#include <webserver/request.h>
#include <navigation/passage.h>
#include <database/books.h>


string navigation_update_url ()
{
  return "navigation/update";
}


bool navigation_update_acl (void * webserver_request)
{
  return Filter_Roles::access_control (webserver_request, Filter_Roles::consultant ());
}


string navigation_update (void * webserver_request)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;

  
  string bible;
  if (request->query.count ("bible")) {
    bible = request->query ["bible"];
  } else {
    bible = request->database_config_user()->getBible ();
  }
  int book = convert_to_int (request->query ["book"]);
  int chapter = convert_to_int (request->query ["chapter"]);
  int verse = convert_to_int (request->query ["verse"]);
  
  
  if (request->query.count ("goback")) {
    Navigation_Passage::goBack (request);
  }
  
  
  else if (request->query.count ("goforward")) {
    Navigation_Passage::goForward (request);
  }
  
  
  else if (request->query.count ("getbooks")) {
    return Navigation_Passage::getBooksFragment (request, bible);
  }
  
  
  else if (request->query.count ("applybook")) {
    string msg = request->query ["applybook"];
    if (msg.find ("cancel") == string::npos) {
      int book = convert_to_int (msg);
      if (book) Navigation_Passage::setBook (request, book);
    }
  }
  
  
  else if (request->query.count ("getchapters")) {
    return Navigation_Passage::getChaptersFragment (request, bible, book, chapter);
  }
  
  
  else if (request->query.count ("applychapter")) {
    string msg = request->query ["applychapter"];
    if (msg.find ("cancel") == string::npos) {
      int chapter = convert_to_int (msg);
      Navigation_Passage::setChapter (request, chapter);
    }
  }
  
  
  else if (request->query.count ("getverses")) {
    return Navigation_Passage::getVersesFragment (request, bible, convert_to_int (book), chapter, verse);
  }
  

  else if (request->query.count ("applyverse")) {
    string msg = request->query ["applyverse"];
    if (msg.find ("cancel") == string::npos) {
      int verse = convert_to_int (msg);
      Navigation_Passage::setVerse (request, verse);
    }
  }
  
  
  /*
  else if (request->query.count ("verse")) {
    int verse = convert_to_int (request->query ["verse"]);
    string book = request->query ["book"];
    if (number_in_string (book) != book) {
      book = convert_to_string (Database_Books::getIdFromEnglish (book));
    }
    int chapter = convert_to_int (request->query ["chapter"]);
    Navigation_Passage::setBookChapterVerse (webserver_request, convert_to_int (book), chapter, verse);
    return "";
  }
   */
  
  
  else if (request->query.count ("previousverse")) {
    Navigation_Passage::gotoPreviousVerse (webserver_request, bible);
  }
  
  
  else if (request->query.count ("nextverse")) {
    Navigation_Passage::gotoNextVerse (webserver_request, bible);
  }
  
  /*
  else if (request->query.count ("passage")) {
    string passage = request->query["passage"];
    Navigation_Passage::setPassage (webserver_request, bible, passage);
    return "";
  }
   */
  
  return Navigation_Passage::getNavigator (request, bible);
}

